db/column_family.cc

```diff
@@ -443,14 +443,23 @@ bool SuperVersion::Unref() {
   return previous_refs == 1;
 }
 
-void SuperVersion::Cleanup() {
+void SuperVersion::Cleanup(const bool fromMemPurge) {
   assert(refs.load(std::memory_order_relaxed) == 0);
+  // Since this SuperVersion object is being deleted,
+  // decrement reference to the immutable MemtableList
+  // this SV object was pointing to.
   imm->Unref(&to_delete);
   MemTable* m = mem->Unref();
   if (m != nullptr) {
-    auto* memory_usage = current->cfd()->imm()->current_memory_usage();
-    assert(*memory_usage >= m->ApproximateMemoryUsage());
-    *memory_usage -= m->ApproximateMemoryUsage();
+    // In MemPurge, the memtable is not made immutable.
+    // Therefore it does not contribute to the imm
+    // memory usage (and actually is not part of
+    // the 'imm' MemtableList at all).
+    if (!fromMemPurge) {
```

Comment:
I'm finding this a bit hard to reason about. At some point, the mutable memtable will be added to the immutable list (if mempurge fails or is ineffective in reclaiming space), right? Can we fix the memory usage accounting in ```SwitchMemtable()``` so this code is not impacted?

Reply:
You are 100% correct, I will fix this by setting the "fromMemPurge" as false if the mempurge fails or is ineffective in reclaiming space. I am also renaming this bool as "noImmMemoryContribution" to make this variable more versatile and explicit.

Reply:
For future reference, I am realizing now that I can remove this flag once the `MemPurge` is made a background task for better performance (next PR), so this is a temporary edit to make the Cleanup code compatible with this MemPurge prototype.

Reply:
```SuperVersion::Cleanup``` is called in the context of the last deref, which may not necessarily be the mempurge context. Can we instead increment ```current->cfd()->imm()->current_memory_usage()``` by the purged memtable size in ```SwitchMemtable```, so the last deref of that memtable doesn't have to worry about the purge status?

Reply:
There is no easy way to increment the `imm()->current_memory_usage()` without adding to the `imm` MemtableList. One option would be to create a new method in the `MemTableList` class, e.g "add_memory_usage", but it is not clear to me that this is a more viable option.